testing/testing_dexterity.rst

```diff
@@ -0,0 +1,188 @@
+
+Inspecting this file, we could see some interesting things:
+
+.. code-block:: python
+
+    from plone.app.testing import TEST_USER_ID
+    from plone.app.testing import setRoles
+    ...
+
+    def test_ct_test_type_adding(self):
+        setRoles(self.portal, TEST_USER_ID, ['Contributor'])
+        obj = api.content.create(
+            container=self.portal, type='TestType', id='test_type'
+        )
+
+        self.assertTrue(
+            ITestType.providedBy(obj),
+            u'ITestType not provided by {0}!'.format(obj.id),
+        )
+
+        # check that deleting the object works too
+        self.assertIn('test_type', self.portal.objectIds())
+        api.content.delete(obj=obj)
+        self.assertNotIn('test_type', self.portal.objectIds())
+
+In this test, for example, we are testing that a contributor can add and delete our content-type.
+
+.. note::
+
+    plone.app.testing gives us a default user for tests. We could import its username in a variable called ``TEST_USER_ID`` and set different roles when needed.
+
+    By default each test is run with that user as logged-in user. If we want to run tests as anonymous users or with a different user, we need to logout/login.
+
+
+Excercise
++++++++++
+
+Try to change the permissions for our content-type and set that only Manager can add it.
+Then fix the test for contributor role and create a new test for the Manager.
+
+..  admonition:: Solution
+    :class: toggle
+
+    In ``rolemap.xml``:
+
+    .. code-block:: xml
+
+        <permission name="plonetraining.testing: Add TestType" acquire="True">
+            <role name="Manager"/>
+        </permission>
+
+    In test case file:
+
+    .. code-block:: python
+
+        from AccessControl.unauthorized import Unauthorized
+        ...
+
+        def test_ct_test_type_adding(self):
+            setRoles(self.portal, TEST_USER_ID, ['Manager'])
+            ...
+
+        def test_ct_test_type_adding_contributor(self):
+            setRoles(self.portal, TEST_USER_ID, ['Contributor'])
+            self.assertRaises(
+                Unauthorized,
+                api.content.create,
+                container=self.portal,
+                type='TestType',
+                id='test_type',
+            )
+
+
+These are ``integration`` tests because we are not testing the browser integration.
+
+We could try to create a ``functional`` test to test how our content-type creation works on browser.
+
+Let's create a new test class in the same file like this:
+
+.. code-block:: python
+
+    from plone.testing.z2 import Browser
+    from from plonetraining.testing.testing import PLONETRAINING_TESTING_FUNCTIONAL_TESTING
+    ...
+
+    class TestTypeFunctionalTest(unittest.TestCase):
+
+        layer = PLONETRAINING_TESTING_FUNCTIONAL_TESTING
+
+        def setUp(self):
```

Comment:
This is the tirst time a `setUp` method is used. It should be explained what they are used for. Same with `tearDown`.

Reply:
done..let me know if it's more clear now.